File: PROG/Week02/Advanced02/Calendar.java
public class Calendar {
	public static AngelaDate readDate() {
		String[] input = In.nextLine().split(" ");
		int d = Integer.parseInt(input[0]);
		int m = Integer.parseInt(input[1]);
		int y = Integer.parseInt(input[2]);

		AngelaDate date = new AngelaDate(d, m, y);

		if (input.length > 3) {
			date.day = input[3].toLowerCase();
		}

		return date;
	}

	/**
	 * This class functions as a container for an Angela date. It handles
	 * calculations between them.
	 */
	public static class AngelaDate {
		int d, m, y;
		// Had to modify the origina Python code a bit because it originally returned a
		// list with mixed datatypes. This allows me to return an AngelaDate that
		// includes both the ints and a string.
		String day;

		public AngelaDate(int d, int m, int y) {
			this.d = d;
			this.m = m;
			this.y = y;
		}

		/**
		 * Subtracts another AngelaDate from this one, and returns the difference in
		 * days. If the other date is in the past, the days will be positive, otherwise
		 * they will be negative.
		 */
		public int subtract(AngelaDate other) {
			int dy = this.y - other.y;
			int dm = this.m - other.m;
			int dd = this.d - other.d;
			dm += dy * 15;
			dd += dm * 30;
			return dd;
		}
	}

	/**
	 * Gets the index of a string in a string array. Wasn't sure how to do it and
	 * didn't want to google it, so here's my suboptimal solution.
	 */
	public static int indexOf(String[] arr, String val) {
		for (int i = 0; i < arr.length; i++) {
			if (arr[i].equals(val))
				return i;
		}
		return -1;
	}

	static String[] days = { "monday", "tuesday", "wednesday", "thursday", "friday" };

	public static void main(String[] args) {
		AngelaDate cur = readDate();
		AngelaDate oth = readDate();
		int day_i = indexOf(days, cur.day);
		int diff = oth.subtract(cur);

		// Fix posible negative indices
		int newDayIndex = (day_i + diff) % 5;
		if (newDayIndex < 0)
			newDayIndex += 5;

		String oth_day = days[newDayIndex];
		String oth_day_cap = oth_day.substring(0, 1).toUpperCase() + oth_day.substring(1);
		System.out.println(oth_day_cap);
	}
}

File: PROG/Week02/Advanced02/Calendar.py
def read_date():
    data = input().split()
    # either there are 3 elements, or the 4th element is a weekday
    data[:3] = [int(i) for i in data[:3]]
    return data


class AngelaDate:
    """This class functions as a container for an Angela date. It handles
    calculations between them."""

    d: int
    m: int
    y: int

    def __init__(self, d, m, y) -> None:
        self.d = d
        self.m = m
        self.y = y

    def subtract(self, other: "AngelaDate"):
        """Subtracts another AngelaDate from this one, and returns the
        difference in days. If the other date is in the past, the days will be
        positive, otherwise they will be negative."""
        dy = self.y - other.y
        dm = self.m - other.m
        dd = self.d - other.d
        dm += dy * 15
        dd += dm * 30
        return dd


days = ["monday", "tuesday", "wednesday", "thursday", "friday"]


def main():
    d1, m1, y1, day = read_date()
    day = day.lower()  # they really did troll me with that one
    d2, m2, y2 = read_date()
    day_i = days.index(day)

    cur = AngelaDate(d1, m1, y1)
    oth = AngelaDate(d2, m2, y2)
    diff = oth.subtract(cur)
    # gets an "absolute" day, then figures where in the 5-day cycle it lands.
    oth_day = days[(day_i + diff) % 5]
    print(day_i)
    print(diff)
    print(oth_day.capitalize())


if __name__ == "__main__":
    main()
